crates/ruff_python_semantic/src/cfg/graph.rs

```diff
@@ -0,0 +1,282 @@
+/// [`Stmt`]s, together with outgoing edges to other basic blocks.
+#[derive(Debug, Default)]
+struct BlockData<'stmt> {
+    kind: BlockKind,
+    stmts: &'stmt [Stmt],
+    out: Edges,
+    parents: SmallVec<[BlockId; 2]>,
+}
+
+impl Ranged for BlockData<'_> {
+    fn range(&self) -> TextRange {
+        let Some(first) = self.stmts.first() else {
+            return TextRange::default();
+        };
+        let Some(last) = self.stmts.last() else {
+            return TextRange::default();
+        };
+
+        TextRange::new(first.start(), last.end())
+    }
+}
+
+#[derive(Debug, Default, Clone, Copy)]
+pub(crate) enum BlockKind {
+    #[default]
+    Generic,
+    /// Entry point of the control flow graph
+    Start,
+    /// Terminal node for the control flow graph
+    Terminal,
+}
+
+/// Holds a collection of edges. Each edge is determined by:
+///  - a [`Condition`] for traversing the edge, and
+///  - a target block, specified by its [`BlockId`].
+#[derive(Debug, Default, Clone)]
+pub struct Edges {
+    conditions: SmallVec<[Condition; 4]>,
+    targets: SmallVec<[BlockId; 4]>,
+}
+
+impl Edges {
+    /// Creates an unconditional edge to the target block
+    fn always(target: BlockId) -> Self {
+        Self {
+            conditions: smallvec![Condition::Always],
+            targets: smallvec![target],
+        }
+    }
+
+    /// Returns iterator over indices of blocks targeted by given edges
+    pub fn targets(&self) -> impl ExactSizeIterator<Item = BlockId> + '_ {
+        self.targets.iter().copied()
+    }
+
+    /// Returns iterator over [`Condition`]s which must be satisfied to traverse corresponding edge
+    pub fn conditions(&self) -> impl ExactSizeIterator<Item = &Condition> {
+        self.conditions.iter()
+    }
+
+    fn is_empty(&self) -> bool {
+        self.targets.is_empty()
+    }
+
+    pub fn filter_targets_by_conditions<'a, T: FnMut(&Condition) -> bool + 'a>(
+        &'a self,
+        mut predicate: T,
+    ) -> impl Iterator<Item = BlockId> + 'a {
+        self.conditions()
+            .zip(self.targets())
+            .filter(move |(cond, _)| predicate(cond))
+            .map(|(_, block)| block)
+    }
+}
+
+/// Represents a condition to be tested in a multi-way branch
+#[derive(Debug, Clone)]
+pub enum Condition {
+    /// Unconditional edge
+    Always,
+}
+
+struct CFGBuilder<'stmt> {
+    /// Control flow graph under construction
+    cfg: CFG<'stmt>,
+    /// Current basic block index
+    current: BlockId,
+    /// Exit block index for current control flow
+    exit: BlockId,
+}
+
+impl<'stmt> CFGBuilder<'stmt> {
+    /// Returns [`CFGBuilder`] with vector of blocks initialized at given capacity and with both initial and terminal blocks populated.
+    fn with_capacity(capacity: usize) -> Self {
+        let mut blocks = IndexVec::with_capacity(capacity);
+        let initial = blocks.push(BlockData {
+            kind: BlockKind::Start,
+            ..BlockData::default()
+        });
```

Comment:
What do you think of always creating an empty block for the start of the CFG. Could this simplify the `set_stmts` and `set_edgs` function because we can create a `add_block(stmts, edges)` function?